src/tool/hpcrun/gpu/gpu-trace.c

```diff
@@ -332,52 +334,54 @@ gpu_trace_stream_id
  void
 )
 {
-  // FIXME: this is a bad way to compute a stream id 
+  // FIXME: this is a bad way to compute a stream id
   int id = 500 + atomic_fetch_add(&stream_id, 1);
 
   return id;
 }
 
 
-static thread_data_t *
+thread_data_t *
 gpu_trace_stream_acquire
 (
  void
 )
 {
-  thread_data_t* td = NULL;
+  thread_data_t *td_self = hpcrun_get_thread_data();
```

Comment:
there seem to be two td instances in play here: td and td_self. why?

Reply:
Keren said that this part of the code, we use a helper thread to set thread-local data for every other thread. Inside **hpcrun_threadMgr_data_get** which is few lines below, we use pthread_setspecific so I believe that without saving td in td_self, we would overwrite it.

Reply:
I understand now. You want to avoid losing the thread data of the helper thread. Rather than than coping with the fact that 

 hpcrun_threadMgr_data_get(id, NULL, &td, true);

has a side effect due to a pthread_setspecific and saving/restoring before/after, wouldn't it be better to have a variant of this routine that doesn't perform the side effect? That might mean 

- one routine that never performs this assignment and its caller is responsible for the side effect when wanted
- two routines - one that does everything except the side effect, and the second, which is a wrapper around the first additionally performs the side effect
- one routine that conditionally performs the side effect based on a boolean argument passed to it

Any of those options sound better than saving/restoring to cope with the side effect.  If the saving/restoring is indeed the best option, it deserves a few lines of comments  in this routine noting that the save/restore is to cope with the side effect.

Reply:
I will check then with Keren and come back to you once we find best soluiton.

Reply:
We can check with @laksono to see if change `hpcrun_threadMgr_data_get` is the best option. If not, I agree that we can add a more comprehensive comment.

Reply:
My preference is the third option: one routine (`hpcrun_threadMgr_data_get`) that conditionally performs the side effect based on an `int option_policy` argument passed to it.
The reason I prefer `int option_policy` is just in case in the future we want to handle another case, and we don't want to add another boolean argument.

Reply:
Do I need to change every call to hpcrun_threadMgr_data_get to wraper function with argument(preserve_self_td=false) ?

Reply:
Is it hpcrun_threadMgr_data_get_safe a good name for wraper?